In [1]:
import csv
import numpy as np

transactions = []

with open("transactions.csv", encoding="utf-8-sig")as f:
    reader = csv.DictReader(f)
    
    for row in reader:
        transactions.append(row)

balance_after = np.array([row["balance_after"] for row in transactions 
                          if row["balance_after"].strip() !=""], dtype=float)

print(balance_after)

[358321.33 506776.75  91532.2  147422.23 394403.86      0.   288363.
 811098.52 315311.22  57036.6  359232.67 677657.62 112718.57 168923.18
 232692.47 425954.41 528079.23 419526.59  78645.55 589330.28 368873.25
 614434.65 561835.04 211374.52  52804.42 476665.23 648542.19 740143.15
 137967.58 457322.44  71986.39 149971.56  97875.54  87895.05 282810.23
 652318.77 200793.64 750273.12 576887.25 653124.77 484805.23 271731.55
  49612.68 203150.7  628224.22  55470.31 610382.59 401285.66 375836.04
 519573.7  128471.42      0.    38567.32 428670.85 322950.41 142724.81
 249807.23 424089.39 302460.52 433536.99 571036.9  459263.8  247471.01
 389614.56 405211.83 537274.15 583418.48 683864.36 344244.9   67385.9
 550486.21  26561.64 658947.89 738241.92 621027.01 352170.58 332954.34
 445993.46 190375.11 384118.89 189926.74 434671.33 571624.58 144246.87
 542554.15 551108.74      0.   150489.35 466209.23 147655.92  77657.73
   8408.85 117517.78 697875.72   6828.11 696436.85 244289.34 512182.67
 799002.3

In [2]:
balances = balance_after

mean_val = balances.mean()
median_val = np.median(balances)

print("==== Central Tendency: balance_after ====")
print(f"Mean:   NGN {mean_val:>10,.2f}")
print(f"Median: NGN {median_val:>10,.2f}")
print()

# Notifying the Gap between the Mean & the Median and what it signals about the data
print(f"Mean - Median Gap:")
print(f"NGN {mean_val - median_val:>10,.2f}")
print(f"DIRECTION: \n  Mean > Median → SLIGHTLY RIGHT-SKEWED (Positive Skewed) distribution")
print(f"INTERPRETATION: \n   The small mean-median gap of ₦1,498.01 suggests that the distribution is only midly right-skewed,\n with a few higher balance values pulling the mean slightly above the median.")
     

==== Central Tendency: balance_after ====
Mean:   NGN 392,473.51
Median: NGN 390,975.50

Mean - Median Gap:
NGN   1,498.01
DIRECTION: 
  Mean > Median → SLIGHTLY RIGHT-SKEWED (Positive Skewed) distribution
INTERPRETATION: 
   The small mean-median gap of ₦1,498.01 suggests that the distribution is only midly right-skewed,
 with a few higher balance values pulling the mean slightly above the median.


In [3]:
# Importing the required libraries and loading the dataset
import pandas as pd
import numpy as np

df = pd.read_csv("transactions.csv", encoding="utf-8-sig")

In [4]:
# Running the standard first-look workflow on the dataset
print(f"RAW DATASET: {df.shape[0]} Rows x {df.shape[1]} Columns")
 

RAW DATASET: 500 Rows x 14 Columns


In [5]:
df.head(10)

,transaction_id,customer_name,phone,email,account_type,branch_state,branch_city,transaction_type,channel,transaction_amount,balance_after,transaction_date,status,narration
0,TXN-0001,Yakubu Musa,8.059518e+09,yakubumusa@hotmail.com,Fixed Deposit,Ogun,Ota,Credit,USSD,₦76458,358321.33,2024-07-10,Successful,Dividend Payment
1,TXN-0002,Yakubu Musa,7.031028e+09,yakubu_musa1@hotmail.com,Savings,Kaduna,Kafanchan,Transfer,Bank Branch,₦1817,506776.75,2024-03-07,Successful,Transfer to Halima Obi - Opay
2,TXN-0003,Yusuf Obi,7.032930e+09,yusufobi@yahoo.com,Current,Delta,Asaba,Transfer,Mobile App,₦17514,91532.20,2024-10-06,Successful,Interbank Transfer - Access Bank
3,TXN-0004,Oluwaseun Okonkwo,8.145667e+09,oluwaseun.okonkwo93@outlook.com,Savings,KADUNA,Kaduna South,Debit,POS,₦5182,147422.23,2024-02-07,Reversed,ATM Cash Withdrawal
4,TXN-0005,Musa Salami,8.166418e+09,musa.salami56@outlook.com,Savings,Rivers,Eleme,Transfer,POS,₦45136,394403.86,2024-12-11,Reversed,NIP Transfer to Tosin Okonkwo
5,TXN-0006,Obinna Salami,8.062464e+09,obinnasalami1@yahoo.com,Savings,DELTA,Sapele,Transfer,Bank Branch,₦125269,0.00,2024-06-21,Failed,Transfer to Chukwuemeka Tanko
6,TXN-0007,Bayo Adebayo,8.024190e+09,bayo_adebayo76@yahoo.com,Current,KADUNA,Kaduna South,Transfer,Bank Branch,₦17580,288363.00,2024-09-28,Pending,NIP Transfer to Babatunde Adekunle
7,TXN-0008,Segun Adebayo,8.134348e+09,segun.adebayo74@yahoo.com,Current,Delta,Ughelli,Credit,Bank Branch,₦193760,811098.52,2024-03-07,Successful,Freelance Income
8,TXN-0009,Chidi Usman,9.065642e+09,chidi_usman30@yahoo.com,Current,RIVERS,Ikwerre,Debit,ATM,₦17021,315311.22,2024-02-06,Failed,ATM Cash Withdrawal
9,TXN-0010,Tosin Bello,8.138112e+09,tosin_bello@hotmail.com,Savings,Delta,Asaba,Debit,Bank Branch,₦24629,57036.60,2024-05-21,Successful,Fuel Payment


In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   transaction_id      500 non-null    str    
 1   customer_name       500 non-null    str    
 2   phone               492 non-null    float64
 3   email               500 non-null    str    
 4   account_type        500 non-null    str    
 5   branch_state        500 non-null    str    
 6   branch_city         500 non-null    str    
 7   transaction_type    500 non-null    str    
 8   channel             500 non-null    str    
 9   transaction_amount  500 non-null    str    
 10  balance_after       494 non-null    float64
 11  transaction_date    500 non-null    str    
 12  status              500 non-null    str    
 13  narration           500 non-null    str    
dtypes: float64(2), str(12)
memory usage: 54.8 KB


In [7]:
df.describe()

,phone,balance_after
count,4.920000e+02,494.000000
mean,8.241655e+09,392473.513644
std,6.085319e+08,233040.603265
min,7.010309e+09,0.000000
25%,8.055629e+09,193352.017500
50%,8.123081e+09,390975.500000
75%,9.026667e+09,581492.760000
max,9.099892e+09,872562.220000


In [8]:
df.isnull().sum()

transaction_id        0
customer_name         0
phone                 8
email                 0
account_type          0
branch_state          0
branch_city           0
transaction_type      0
channel               0
transaction_amount    0
balance_after         6
transaction_date      0
status                0
narration             0
dtype: int64

### QUALITY PROBLEMS IDENTIFIED FROM THE DATASET OUTPUTS

1. There are missing values in the **phone** column (8 Missing) and in the **balance_after** column (6 Missing) that need to be treated.

2. The **transaction_amount** is in the wrong datatype format; it should be a calculated number (float) not to be treated as a string, because of currency symbols like "₦".

3. Also, the **transaction_date** is in the wrong datatype format too and should be treated as a datetime, not a string.

4. In the **branch_state** column, there is inconsistency in their naming and capitalization (such as "Kaduna" vs. "KADUNA").

5. The **phone** column is in the wrong datatype format as well; it should be stored as a string instead of a float so it does not display in scientific notation.

##### CLEANING THE DATA

In [9]:
# WORKING ON A COPY

df_clean = df.copy()

print("Working Copy Created:", df_clean.shape)
print("Original Dataset:    ", df.shape)

Working Copy Created: (500, 14)
Original Dataset:     (500, 14)


In [10]:
# Dealing with Missing Values by filling with Not Provided

df_clean["phone"] = df_clean["phone"].fillna("Not Provided")

print("Missing Phone Values Filled:", df_clean["phone"].isnull().sum())
print("Sample phone values after filling missing values:")
print(df_clean["phone"].value_counts().head())

Missing Phone Values Filled: 0
Sample phone values after filling missing values:
phone
Not Provided    8
9096948779.0    2
8085720802.0    2
8050033287.0    2
8076470917.0    2
Name: count, dtype: int64


In [11]:
# Changing the Phone Column to String Type

df["phone"] = df["phone"].astype("Int64").astype(str)

df.head(3)

,transaction_id,customer_name,phone,email,account_type,branch_state,branch_city,transaction_type,channel,transaction_amount,balance_after,transaction_date,status,narration
0,TXN-0001,Yakubu Musa,8059518175,yakubumusa@hotmail.com,Fixed Deposit,Ogun,Ota,Credit,USSD,₦76458,358321.33,2024-07-10,Successful,Dividend Payment
1,TXN-0002,Yakubu Musa,7031027616,yakubu_musa1@hotmail.com,Savings,Kaduna,Kafanchan,Transfer,Bank Branch,₦1817,506776.75,2024-03-07,Successful,Transfer to Halima Obi - Opay
2,TXN-0003,Yusuf Obi,7032929611,yusufobi@yahoo.com,Current,Delta,Asaba,Transfer,Mobile App,₦17514,91532.20,2024-10-06,Successful,Interbank Transfer - Access Bank


### **Method Specified:**
Impute missing phone numbers with the placeholder string 'Not Provided'.

### **Technical Justification:**
Phone numbers represent unique Personal Identifiable Information (PII) that cannot be accurately estimated or imputed statistically. Dropping these rows would result in unnecessary data loss for non-contact-related analysis. Using 'Not Provided' preserves the transaction record while maintaining data integrity.

In [12]:
df_clean.isnull().sum()

transaction_id        0
customer_name         0
phone                 0
email                 0
account_type          0
branch_state          0
branch_city           0
transaction_type      0
channel               0
transaction_amount    0
balance_after         6
transaction_date      0
status                0
narration             0
dtype: int64

In [13]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   transaction_id      500 non-null    str    
 1   customer_name       500 non-null    str    
 2   phone               500 non-null    object 
 3   email               500 non-null    str    
 4   account_type        500 non-null    str    
 5   branch_state        500 non-null    str    
 6   branch_city         500 non-null    str    
 7   transaction_type    500 non-null    str    
 8   channel             500 non-null    str    
 9   transaction_amount  500 non-null    str    
 10  balance_after       494 non-null    float64
 11  transaction_date    500 non-null    str    
 12  status              500 non-null    str    
 13  narration           500 non-null    str    
dtypes: float64(1), object(1), str(12)
memory usage: 54.8+ KB


In [14]:
# Dropping rows with missing values in the 'balance_after' column

before_drop_shape = df_clean.shape
df_clean = df_clean.dropna(subset=["balance_after"])
after_drop_shape = df_clean.shape

print(f"Before Drop Null Values: {before_drop_shape[0]} Rows")
print(f"After Drop Null Values:   {after_drop_shape[0]} Rows")
print(f"Null Values Dropped:        {before_drop_shape[0] - after_drop_shape[0]}")
print(f"Missing balance_after Values: {df_clean['balance_after'].isnull().sum()}")

Before Drop Null Values: 500 Rows
After Drop Null Values:   494 Rows
Null Values Dropped:        6
Missing balance_after Values: 0


### **Method Specified:**
Drop rows containing missing values in the balance_assets column.

### **Technical Justification:**
Asset balances are strict financial metrics that should not be estimated using statistical imputation (e.g., mean or median), as this introduces artificial variance into financial calculations. Dropping these records ensures that balance aggregations reflect only true, verified balances.

In [15]:
df_clean.isnull().sum()

transaction_id        0
customer_name         0
phone                 0
email                 0
account_type          0
branch_state          0
branch_city           0
transaction_type      0
channel               0
transaction_amount    0
balance_after         0
transaction_date      0
status                0
narration             0
dtype: int64

In [16]:
df_clean.info()

<class 'pandas.DataFrame'>
Index: 494 entries, 0 to 499
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   transaction_id      494 non-null    str    
 1   customer_name       494 non-null    str    
 2   phone               494 non-null    object 
 3   email               494 non-null    str    
 4   account_type        494 non-null    str    
 5   branch_state        494 non-null    str    
 6   branch_city         494 non-null    str    
 7   transaction_type    494 non-null    str    
 8   channel             494 non-null    str    
 9   transaction_amount  494 non-null    str    
 10  balance_after       494 non-null    float64
 11  transaction_date    494 non-null    str    
 12  status              494 non-null    str    
 13  narration           494 non-null    str    
dtypes: float64(1), object(1), str(12)
memory usage: 57.9+ KB


In [17]:
# TO CHECK DUPLICATE ROWS IN THE DATASET

print(f"Duplicate Rows: {df_clean.duplicated().sum()}")
print()
print("THE DUPLICATE ROWS:")
df[df.duplicated(keep=False)].sort_values("balance_after")

Duplicate Rows: 4

THE DUPLICATE ROWS:


,transaction_id,customer_name,phone,email,account_type,branch_state,branch_city,transaction_type,channel,transaction_amount,balance_after,transaction_date,status,narration
495,TXN-0496,Halima Nwosu,8076470917,halima.nwosu7@gmail.com,Savings,rivers,Obio-Akpor,Credit,USSD,₦11434,29638.93,2024-05-25,Reversed,Freelance Income
499,TXN-0496,Halima Nwosu,8076470917,halima.nwosu7@gmail.com,Savings,rivers,Obio-Akpor,Credit,USSD,₦11434,29638.93,2024-05-25,Reversed,Freelance Income
493,TXN-0494,Ngozi Okafor,8085720802,ngozi_okafor24@hotmail.com,Savings,Delta,Warri,Transfer,ATM,₦5980,243429.78,2024-06-16,Pending,Transfer to Chinyere Okafor - Opay
497,TXN-0494,Ngozi Okafor,8085720802,ngozi_okafor24@hotmail.com,Savings,Delta,Warri,Transfer,ATM,₦5980,243429.78,2024-06-16,Pending,Transfer to Chinyere Okafor - Opay
494,TXN-0495,Ifeanyi Nwofor,8050033287,ifeanyi.nwofor72@yahoo.com,Savings,lagos,Victoria Island,Transfer,POS,₦6924,288763.59,2024-10-29,Failed,Transfer to Zainab Bello - Opay
498,TXN-0495,Ifeanyi Nwofor,8050033287,ifeanyi.nwofor72@yahoo.com,Savings,lagos,Victoria Island,Transfer,POS,₦6924,288763.59,2024-10-29,Failed,Transfer to Zainab Bello - Opay
492,TXN-0493,Kayode Adeyemi,9096948779,kayodeadeyemi50@outlook.com,Savings,KANO,Fagge,Transfer,Mobile App,₦2631,544813.86,2024-05-31,Reversed,Interbank Transfer - Access Bank
496,TXN-0493,Kayode Adeyemi,9096948779,kayodeadeyemi50@outlook.com,Savings,KANO,Fagge,Transfer,Mobile App,₦2631,544813.86,2024-05-31,Reversed,Interbank Transfer - Access Bank


In [18]:
# To Remove Duplicate Rows in the Dataset

before_drop_shape = df_clean.shape
df_clean = df_clean.drop_duplicates()
after_drop_shape = df_clean.shape

print(f"Before Drop Duplicates: {before_drop_shape[0]} Rows")
print(f"After Drop Duplicates:   {after_drop_shape[0]} Rows")   
print(f"Duplicate Rows Dropped:  {before_drop_shape[0] - after_drop_shape[0]}")

Before Drop Duplicates: 494 Rows
After Drop Duplicates:   490 Rows
Duplicate Rows Dropped:  4


### **Method Specified:**
Identify and drop duplicate rows across the dataset.

### **Technical Justification:**
Duplicate entries artificially inflate transaction volumes and distort summary metrics. Removing redundant rows ensures that each transaction is uniquely counted, guaranteeing accurate downstream aggregations.

In [19]:
df_clean.info()

<class 'pandas.DataFrame'>
Index: 490 entries, 0 to 495
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   transaction_id      490 non-null    str    
 1   customer_name       490 non-null    str    
 2   phone               490 non-null    object 
 3   email               490 non-null    str    
 4   account_type        490 non-null    str    
 5   branch_state        490 non-null    str    
 6   branch_city         490 non-null    str    
 7   transaction_type    490 non-null    str    
 8   channel             490 non-null    str    
 9   transaction_amount  490 non-null    str    
 10  balance_after       490 non-null    float64
 11  transaction_date    490 non-null    str    
 12  status              490 non-null    str    
 13  narration           490 non-null    str    
dtypes: float64(1), object(1), str(12)
memory usage: 57.4+ KB


In [20]:
# TO STRIP THE ₦ SYMBOL FROM THE TRANSACTION_AMOUNT COLUMNS AND CONVERT TO FLOAT.
print("Before — dtype:", df_clean["transaction_amount"].dtype)
print("Before — sample:", df_clean["transaction_amount"].head(5).tolist())

# Step 1: remove the ₦ symbol
df_clean["transaction_amount"] = df_clean["transaction_amount"].str.replace("₦", "", regex=False)

# Step 2: convert to numeric (errors='coerce')
df_clean["transaction_amount"] = pd.to_numeric(df_clean["transaction_amount"], errors='coerce')

print()
print("After — dtype:", df_clean["transaction_amount"].dtype)
print("After — sample:", df_clean["transaction_amount"].head(5).tolist())
print("Find Any Missing New Values?", df_clean["transaction_amount"].isnull().sum())

Before — dtype: str
Before — sample: ['₦76458', '₦1817', '₦17514', '₦5182', '₦45136']

After — dtype: int64
After — sample: [76458, 1817, 17514, 5182, 45136]
Find Any Missing New Values? 0


### **Method Specified:**
Strip the currency symbol (₦) and commas, then convert the column data type from string to numeric (float/int).

### **Technical Justification:**
Financial values formatted as strings with currency symbols block mathematical operations and statistical aggregation. Removing non-numeric characters and casting the column to a numeric data type enables numerical computations, KPI calculations, and plotting.

In [21]:
# TO STANDARDISE BRANCH_STATE TO TITLE CASE
print("Before - unique branch_state value:")
print(sorted(df_clean["branch_state"].unique()))
print(f"Total Unique Values: {df_clean['branch_state'].nunique()}")

df_clean["branch_state"] = df_clean["branch_state"].str.strip().str.title()

print()
print()
print("After - unique branch_state value:")
print(sorted(df_clean["branch_state"].unique()))
print(f"Total Unique Values: {df_clean['branch_state'].nunique()}")

Before - unique branch_state value:
['ABUJA', 'ANAMBRA', 'Abuja', 'Anambra', 'DELTA', 'Delta', 'ENUGU', 'Enugu', 'KADUNA', 'KANO', 'Kaduna', 'Kano', 'LAGOS', 'Lagos', 'OGUN', 'OYO', 'Ogun', 'Oyo', 'RIVERS', 'Rivers', 'abuja', 'anambra', 'delta', 'enugu', 'kaduna', 'kano', 'lagos', 'ogun', 'oyo', 'rivers']
Total Unique Values: 30


After - unique branch_state value:
['Abuja', 'Anambra', 'Delta', 'Enugu', 'Kaduna', 'Kano', 'Lagos', 'Ogun', 'Oyo', 'Rivers']
Total Unique Values: 10


### **Method Specified:**
Trim leading/trailing whitespace and apply Title Case formatting to the branch names.

### **Technical Justification:**
Variations in casing (e.g., "ANAMBRA" vs "anambra") and hidden spaces cause data processing engines to treat identical locations as separate categories. Standardizing string formatting guarantees accurate categorical grouping during GROUP BY operations and visualization.

In [22]:
# TO CONVERT TRANSACTION_DATE TO DATETIME 
print("Before — dtype:", df_clean["transaction_date"].dtype)
print("Before — sample:", df_clean["transaction_date"].head(3).tolist())

df_clean["transaction_date"] = pd.to_datetime(df_clean["transaction_date"])

print()
print()
print("After — dtype:", df_clean["transaction_date"].dtype)
print("After — sample:", df_clean["transaction_date"].head(3).tolist())

Before — dtype: str
Before — sample: ['2024-07-10', '2024-03-07', '2024-10-06']


After — dtype: datetime64[us]
After — sample: [Timestamp('2024-07-10 00:00:00'), Timestamp('2024-03-07 00:00:00'), Timestamp('2024-10-06 00:00:00')]


In [23]:
# TO EXTRACT MONTH COLUMN FROM TRANSACTION_DATE

df_clean["transaction_month"] = df_clean["transaction_date"].dt.month

print("Transaction Per Month:")
print(df_clean["transaction_month"].value_counts().sort_index())


Transaction Per Month:
transaction_month
1     45
2     40
3     40
4     47
5     44
6     31
7     37
8     38
9     44
10    45
11    35
12    44
Name: count, dtype: int64


### **Method Specified:**
Convert transaction_date to datetime format and extract the month into a separate column.

### **Technical Justification:**
String-based dates cannot be sorted sequentially or grouped by time intervals. Parsing to datetime enables time-series tracking and allows the extraction of time attributes (such as month), which supports temporal trends analysis and dashboard filtering.

In [24]:
df_clean.head(10)

,transaction_id,customer_name,phone,email,account_type,branch_state,branch_city,transaction_type,channel,transaction_amount,balance_after,transaction_date,status,narration,transaction_month
0,TXN-0001,Yakubu Musa,8059518175.0,yakubumusa@hotmail.com,Fixed Deposit,Ogun,Ota,Credit,USSD,76458,358321.33,2024-07-10,Successful,Dividend Payment,7
1,TXN-0002,Yakubu Musa,7031027616.0,yakubu_musa1@hotmail.com,Savings,Kaduna,Kafanchan,Transfer,Bank Branch,1817,506776.75,2024-03-07,Successful,Transfer to Halima Obi - Opay,3
2,TXN-0003,Yusuf Obi,7032929611.0,yusufobi@yahoo.com,Current,Delta,Asaba,Transfer,Mobile App,17514,91532.20,2024-10-06,Successful,Interbank Transfer - Access Bank,10
3,TXN-0004,Oluwaseun Okonkwo,8145666581.0,oluwaseun.okonkwo93@outlook.com,Savings,Kaduna,Kaduna South,Debit,POS,5182,147422.23,2024-02-07,Reversed,ATM Cash Withdrawal,2
4,TXN-0005,Musa Salami,8166418457.0,musa.salami56@outlook.com,Savings,Rivers,Eleme,Transfer,POS,45136,394403.86,2024-12-11,Reversed,NIP Transfer to Tosin Okonkwo,12
5,TXN-0006,Obinna Salami,8062463656.0,obinnasalami1@yahoo.com,Savings,Delta,Sapele,Transfer,Bank Branch,125269,0.00,2024-06-21,Failed,Transfer to Chukwuemeka Tanko,6
6,TXN-0007,Bayo Adebayo,8024190064.0,bayo_adebayo76@yahoo.com,Current,Kaduna,Kaduna South,Transfer,Bank Branch,17580,288363.00,2024-09-28,Pending,NIP Transfer to Babatunde Adekunle,9
7,TXN-0008,Segun Adebayo,8134347962.0,segun.adebayo74@yahoo.com,Current,Delta,Ughelli,Credit,Bank Branch,193760,811098.52,2024-03-07,Successful,Freelance Income,3
8,TXN-0009,Chidi Usman,9065642290.0,chidi_usman30@yahoo.com,Current,Rivers,Ikwerre,Debit,ATM,17021,315311.22,2024-02-06,Failed,ATM Cash Withdrawal,2
9,TXN-0010,Tosin Bello,8138112228.0,tosin_bello@hotmail.com,Savings,Delta,Asaba,Debit,Bank Branch,24629,57036.60,2024-05-21,Successful,Fuel Payment,5


In [25]:
# FULL VALIDATION BLOCK CONFIRMATION
print("======== VALIDATION REPORT ========")
print()

print(f"1. Original Shape: {df.shape}")
print(f"   Cleaned Shape:  {df_clean.shape}")

print(f"2. Rows removed:   {df.shape[0] - df_clean.shape[0]}")

print()

print(f"3. Original Duplicates: {df.duplicated().sum()}")
print(f"   Cleaned Duplicates:  {df_clean.duplicated().sum()}")

print()

print("4. Unique Branch States:")
print(df_clean["branch_state"].unique())

print()

print("5. Original Info:")
df.info()

print("\nCleaned Info:")
df_clean.info()

======== VALIDATION REPORT ========

1. Original Shape: (500, 14)
   Cleaned Shape:  (490, 15)
2. Rows removed:   10

3. Original Duplicates: 4
   Cleaned Duplicates:  0

4. Unique Branch States:
<StringArray>
[   'Ogun',  'Kaduna',   'Delta',  'Rivers',   'Enugu',   'Lagos',   'Abuja',
 'Anambra',    'Kano',     'Oyo']
Length: 10, dtype: str

5. Original Info:
<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   transaction_id      500 non-null    str    
 1   customer_name       500 non-null    str    
 2   phone               492 non-null    str    
 3   email               500 non-null    str    
 4   account_type        500 non-null    str    
 5   branch_state        500 non-null    str    
 6   branch_city         500 non-null    str    
 7   transaction_type    500 non-null    str    
 8   channel             500 non-null    str    
 9 

##### TRANSFORM & ANSWER BUSINESS QUESTIONS

In [26]:
# USING GROUPBY.AGG TO ANSWER THE QUESTIONS

# Step 1: Group by state and calculate the three metrics
state_summary = df_clean.groupby("branch_state").agg(
    total_volume=("transaction_amount", "sum"),
    order_count=("transaction_amount", "count"),
    avg_transaction=("transaction_amount", "mean")
)

In [27]:
# Question 1: Branch State Summary Table
state_summary = df_clean.groupby("branch_state").agg(
    total_volume=("transaction_amount", "sum"),
    order_count=("transaction_amount", "count"),
    avg_transaction=("transaction_amount", "mean")
).sort_values(by="total_volume", ascending=False)

# Display the summary table
state_summary.head(10)

,total_volume,order_count,avg_transaction
branch_state,,,
Ogun,5117847,57,89786.789474
Delta,4876964,56,87088.642857
Kaduna,3095498,46,67293.434783
Anambra,2252352,50,45047.040000
Rivers,2062416,48,42967.000000
Abuja,1880916,49,38386.040816
Oyo,1665026,46,36196.217391
Enugu,1604737,49,32749.734694
Lagos,1104352,46,24007.652174


##### INTERPRETING THE BRANCH STATE ANALYSIS FOR BUSINESS
1. Top Revenue Generator: Ogun State ranks highest in total transaction volume at $\text{₦}51,178,475.79$ across 57 transactions, achieving the highest average order value ($\text{₦}897,867.89$) among all states.

2. Volume Leaders: Delta ($\text{₦}48.77\text{M}$) and Kaduna ($\text{₦}30.95\text{M}$) follow as key volume drivers.

3. Business Impact: Ogun and Delta are high-value markets for the bank. Operations should prioritize credit line availability and high-tier customer support services in these branches to retain high-value account holders.

In [28]:
# Question 2: Channel Summary Table
channel_summary = df_clean.groupby("channel").agg(
    frequency=("transaction_amount", "count"),
    avg_transaction=("transaction_amount", "mean")
).sort_values(by="frequency", ascending=False)

# Display the summary table
channel_summary.head(10)

,frequency,avg_transaction
channel,,
POS,110,64569.245455
Bank Branch,103,34301.747573
ATM,99,70608.242424
USSD,98,40089.642857
Mobile App,80,39124.112500


##### INTERPRETING THE CHANNEL PERFORMANCE ANALYSIS FOR BUSINESS
1. Dominant Channel: POS is both the most frequently used channel (110 transactions) and yields the highest average transaction amount ($\text{₦}645,699.25$).

2. Secondary Channels: Bank Branch remains popular for high-value operations (103 transactions, averaging $\text{₦}343,017.75$), while Mobile App handles smaller, lower-value transactions (80 transactions, averaging $\text{₦}391,241.11$).

3. Business Impact: Since POS handles the largest throughput, technical infrastructure for POS transaction processing must remain reliable to prevent revenue loss. The bank can also run campaigns to convert Mobile App users into higher-volume transactions.
